# 2.0 Intro to Computer Vision

References

> AI and Machine Learning for On Device Development - Laurence Moroney

> * https://www.oreilly.com/library/view/ai-and-machine/9781098101732/

> * https://github.com/lmoroney/odmlbook/tree/main/BookSource

> Fashion MNIST

> * https://arxiv.org/abs/1708.07747

> * https://paperswithcode.com/dataset/fashion-mnist

> Keras

> * https://keras.io/

> Tensorflow

> * https://www.tensorflow.org/


IDE (Interactive Development Environment)

>[Colab](https://colab.research.google.com)

We are going to look at a script that uses computer vision to classify clothing

> * Fashion MNIST, a new dataset comprising of 28x28 grayscale images of 70,000 fashion products from 10 categories, with 7,000 images per category. The training set has 60,000 images and the test set has 10,000 images. Fashion-MNIST is intended to serve as a direct drop-in replacement for the original MNIST dataset for benchmarking machine learning algorithms, as it shares the same image size, data format and the structure of training and testing splits.

In [1]:
import tensorflow as tf

data = tf.keras.datasets.mnist
(training_images, training_labels), (val_images, val_labels) = data.load_data()

training_images  = training_images / 255.0
val_images = val_images / 255.0

model = tf.keras.models.Sequential(
[tf.keras.layers.Flatten(input_shape=(28,28)),
 tf.keras.layers.Dense(20, activation=tf.nn.relu),
     tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(training_images, training_labels, epochs=20)

model.evaluate(val_images, val_labels)

11490434/11490434 [==============================] - 1s 0us/step
Epoch 1/20
1875/1875 [==============================] - 11s 5ms/step - loss: 0.4036 - accuracy: 0.8871
Epoch 2/20
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2318 - accuracy: 0.9334
Epoch 3/20
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1952 - accuracy: 0.9444
Epoch 4/20
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1719 - accuracy: 0.9503
Epoch 5/20
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1532 - accuracy: 0.9554
Epoch 6/20
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1406 - accuracy: 0.9591
Epoch 7/20
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1303 - accuracy: 0.9620
Epoch 8/20
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1222 - accuracy: 0.9646
Epoch 9/20
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1164 - accuracy: 0.9653
Epoch 10/20
1875

[0.14016012847423553, 0.9607999920845032]

Let's take a look at each of the lines of code

This line imports the python libraries that we will need to run the script

> https://www.tensorflow.org/

In [ ]:
import tensorflow as tf

This line loads the Fashion MNIST database, which we will evaluate, via tensorflow and keras

In [ ]:
data = tf.keras.datasets.mnist
(training_images, training_labels), (val_images, val_labels) = data.load_data()

This line scales down the pixels in the 60,000 training images, and the process is called normalization

> Pixels normally have a range of values from 0 to 255

> We are scaling that value range down to 0 to 1

> A smaller pixel value range works well with the activation functions that we use in the neurons (nodes)

> * errors do not massively inflate when using smaller pixel values

In [ ]:
training_images  = training_images / 255.0

This line scales down the pixels in the 10,000 validation (test) images, and the process is called normalization

> Pixels normally have a range of values from 0 to 255

> We are scaling that range down to 0 to 1

> A smaller pixel range works well with the activation functions that we use in the neurons (nodes)

> * errors do not massively inflate when using smaller pixel values

In [ ]:
val_images = val_images / 255.0

This line defines the model architecture

> Sequential()

> * A Sequential model is appropriate for a plain stack of layers where each layer has exactly one input tensor and one output tensor

> * https://www.tensorflow.org/api_docs/python/tf/keras/Sequential

> * https://www.tensorflow.org/api_docs/python/tf/keras/Sequential

> .Flatten()

> * Flattens the input. Does not affect the batch size.

> * https://keras.io/api/layers/reshaping_layers/flatten/

> .Dense()

> * Just your regular densely-connected NN layer

> * https://keras.io/api/layers/core_layers/dense/

> nn.relu

> * activation function - mimics the firing thresholds of a biological neuron

> * https://www.tensorflow.org/api_docs/python/tf/nn/relu

> nn.softmax

> * activation function - mimics the firing thresholds of a biological neuron

> * https://www.tensorflow.org/api_docs/python/tf/nn/softmax

In [ ]:
model = tf.keras.models.Sequential(
[tf.keras.layers.Flatten(input_shape=(28,28)),
 tf.keras.layers.Dense(20, activation=tf.nn.relu),
     tf.keras.layers.Dense(10, activation=tf.nn.softmax)])

This line configures the model for training

> .compile()

> * https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile

> optimizer - algorithm selections

> * https://www.tensorflow.org/api_docs/python/tf/keras/optimizers

> loss - between true labels and predicted labels

> * https://www.tensorflow.org/api_docs/python/tf/keras/losses

> metrics - List of various metrics to be evaluated by the model during training and testing

> * https://www.tensorflow.org/api_docs/python/tf/keras/metrics

In [ ]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

This line trains the model for a fixed number of epochs (dataset iterations) using the .fit method

> * https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit

> * https://keras.io/api/models/model_training_apis/

> * a loop that makes a guess, measures how good or how bad that loss is, optimizes for a new guess, and repeats

> *  epochs = 20, indicates that we’ll repeat the loop 20 times

> TensorFlow supports batching to make things faster

> * Fashion MNIST defaults to each batch having 32 images, so it trains with one batch of images at a time

> * This gives you 1875 batches of images to make up 60,000 images (i.e., 60,000 divided by 32 = 1875)


In [ ]:
model.fit(training_images, training_labels, epochs=20)

This line passes the 10,000 validation (test) images to the model to see how it accurately it parses (classifies, labels) them

> * 313/313 [==============================] - 1s 4ms/step - loss: 0.1402 - accuracy: 0.9608
[0.14016012847423553, 0.9607999920845032]

> * our output reflects a 96% classification (labeling) accuracy with respect to assigning clothing labels to clothing images that it has not seen before

> * we should always keep in mind the mathematical concept of overfitting as we evaulate our inputs and outputs ... GIGO

In [ ]:
model.evaluate(val_images, val_labels)